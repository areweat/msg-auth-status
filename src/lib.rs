#![warn(
    clippy::unwrap_used,
//    missing_docs,
    rust_2018_idioms,
    unused_lifetimes,
    unused_qualifications
)]
#![doc = include_str!("../README.md")]

mod types;
pub use types::{AuthenticationResults, HostVersion, Prop};

pub mod auth;
pub mod dkim;
pub mod iprev;
pub mod spf;

mod parser;

#[derive(Debug)]
pub struct MessageAuthStatus<'hdr> {
    results: Vec<AuthenticationResults<'hdr>>,
}

#[derive(Debug)]
pub enum Error {
    ParseNone,
}

impl<'hdr> MessageAuthStatus<'hdr> {
    #[cfg(feature = "mail_parser")]
    pub fn from_mail_parser(msg: &'hdr mail_parser::Message<'hdr>) -> Result<Self, Error> {
        let mut new_self = Self { results: vec![] };

        new_self.results = msg
            .header_values("Authentication-Results")
            .into_iter()
            .map(|mh| mh.try_into().unwrap())
            .collect();

        Ok(new_self)
    }
}

#[cfg(test)]
mod test {
    #[allow(unused_imports)]
    use super::*;
    #[allow(unused_imports)]
    use insta::assert_debug_snapshot;
    #[allow(unused_imports)]
    use rstest::rstest;
    #[allow(unused_imports)]
    use std::{fs::File, io::Read, path::PathBuf};

    #[allow(dead_code)]
    fn load_test_data(file_location: &str) -> Vec<u8> {
        let mut file = File::open(file_location).unwrap();
        let mut data: Vec<u8> = vec![];
        file.read_to_end(&mut data).unwrap();
        data
    }

    #[rstest]
    #[cfg(feature = "mail_parser")]
    fn from_mail_parser(#[files("test_data/rfc8601_b*.txt")] file_path: PathBuf) {
        // set_snapshot_path

        let new_snapshot_path = file_path.with_extension("snap");
        //panic!("snapshot_path = {:?}", new_snapshot_path);

        insta::with_settings!({snapshot_path => new_snapshot_path}, {
            insta::allow_duplicates! {
                let data = load_test_data(file_path.to_str().unwrap());

                let parser = mail_parser::MessageParser::default();

                let parsed = parser.parse(&data).unwrap();

                let status = MessageAuthStatus::from_mail_parser(&parsed);
                assert_debug_snapshot!(&status);
            }
        });
    }
}
