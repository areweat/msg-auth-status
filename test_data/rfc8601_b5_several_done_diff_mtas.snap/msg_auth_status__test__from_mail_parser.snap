---
source: src/lib.rs
expression: "&status"
---
Ok(
    MessageAuthStatus {
        results: [
            AuthenticationResults {
                host: Some(
                    HostVersion {
                        host: "example.com",
                        version: None,
                    },
                ),
                smtp_auth_result: [],
                spf_result: [],
                dkim_result: [
                    DkimResult {
                        code: Pass,
                        reason: None,
                        header_d: Some(
                            "example.com",
                        ),
                        header_i: None,
                        header_b: None,
                        header_a: None,
                        header_s: None,
                        raw: Some(
                            "dkim=pass (good signature) header.d=example.com",
                        ),
                    },
                ],
                iprev_result: [],
                none_done: false,
                raw: Some(
                    "example.com;\n          dkim=pass (good signature) header.d=example.com",
                ),
            },
            AuthenticationResults {
                host: Some(
                    HostVersion {
                        host: "example.com",
                        version: None,
                    },
                ),
                smtp_auth_result: [
                    SmtpAuthResult {
                        code: Pass,
                        smtp_auth: Some(
                            "sender@example.com",
                        ),
                        smtp_mailfrom: None,
                        raw: Some(
                            "auth=pass (cram-md5) smtp.auth=sender@example.com",
                        ),
                    },
                ],
                spf_result: [
                    SpfResult {
                        code: Fail,
                        reason: None,
                        smtp_mailfrom: Some(
                            "example.com",
                        ),
                        smtp_helo: None,
                        raw: Some(
                            "spf=fail smtp.mailfrom=example.com",
                        ),
                    },
                ],
                dkim_result: [],
                iprev_result: [],
                none_done: false,
                raw: Some(
                    "example.com;\n          auth=pass (cram-md5) smtp.auth=sender@example.com;\n          spf=fail smtp.mailfrom=example.com",
                ),
            },
        ],
    },
)
